ci/pipelines/s3-resource-simple-template.yml

```diff
@@ -92,51 +78,25 @@ resources:
     region_name: eu-west-1
     access_key_id: ((concourse_user_access_key))
     secret_access_key: ((concourse_user_secret_key))
+    # For this particular pipeline we use the /artifacts paths also for the master branch,
+    # so that the files uploaded by the tests will expire and be removed periodically.
     path: artifacts/s3-resource-simple/((branch))/
-    sync: false # Disable downloading.
-    options:
-    # FIXME! `--recursive` is _currently_ needed to copy a directory :-(
-    - --recursive
+    # DANGER DANGER DANGER
+    # For vast the majority of your use cases, DO NOT enable downloading (do NOT set sync: true)!
+    # You will download, on each get (explicit and implicit) EVERYTHING contained in the bucket!
+    # This resource has a different behavior than the s3 resource!
+    # We enable it only because this pipeline is special: it is testing the resource itself.
+    sync: true # DANGER - DO NOT COPY THIS LINE - DANGER
+    put_options:
```

Comment:
In my understanding this is a breaking change, that could potentially break a lot of other pipelines?

Reply:
Example: https://github.com/Pix4D/pix4d-calibration/blob/3166f00f4d9aa9e23955143ea54c8b55cfaba3ad/ci/pipelines/pix4d-calibration-template.yml#L381

Note that the number of pipelines like that is quite large (all pipelines: both feature and master: 107)

Reply:
the link to pix4d-calibration-template.yml is:
```
source:
    <<: *s3-params
    region: eu-west-1  # region key name differs from resource `s3` region key name...
    path: final/((project))/((branch))/user-guide
    sync: false # Disable downloading.
    options:
      # FIXME! `--recursive` is _currently_ needed to copy a directory :-(
      - --recursive
```

This PR does not change this, see the file "out" in this PR.

Reply:
But that means you can add `--recursive` to both `options` and the new `put_options` so one can have?
```
    options:
      - --recursive    
    put_options:
      - --recursive
```

Reply:
I am not sure I understand the intent of the question.
To answer the question directly, yes, that example will fail on put.
Both options and put_options are pass-through to the underlying "aws s3".

Reply:
Just for my understanding: isn't it better to leave the put step as it is and actually remove the options from get step
or some tricks with: `get_params:` in put step if needed.

Why? put step is actually the only step used across all other pipelines. Get step is basically used only in this pipeline.

Reply:
Let's take a step back.

The intent of this PR is to be able to test automatically that the put step:
1. uploads the files requested
2. is able to upload also directories and its contents.

To upload recursively, before this PR, the resource needed:
```
options:
  - --recursive    
```
Now to the test. The test must be done in the get step. To tell the resource to actually download something, the resource needs:
```
sync: true
```

In the get step, if "sync: false" (the default), the aws CLI is not called.
If on the other hand "sync: true", then the aws CLI is called as "aws s3 sync ... $options.
This means that asw s3 sync will see also --recursive, which is valid for cp but not for sync, and error out.

To allow testing, this PR allows to pass --recursive (actually anything) as put_option, so that aws sync (the get step) will not see it.

Existing pipelines will not be affected (unless I made an error in the reasoning).
New pipelines should use put_option, because it is more explicit.

Is there something I missed?

Reply:
What I mean is that you don't need to call `aws s3 sync` to implement the `get` step. Calling `aws s3 cp --recursive` does the same thing. The only difference is that you can pass `--delete` to `aws s3 sync` and it will remove local files that are not present remotely (for the `get` step).

Reply:
@odormond I wrote my reply above without seeing yours, this might explain some of the disconnect :-).

The history of this resource is complicated: it is a fork, somebody else from Pix4D did some changes, more than once, and we are now here keeping it alive.

This PR is only meant to allow automatic testing of the default behavior.

Let's have a face to face chat, it will be easier to reach a common understanding.